src/cmake.ts

```diff
@@ -422,6 +421,453 @@ export class ConfigurationReader {
+        const document = this.getTextDocumentByFileName(file);
+        if (document && (line < document.lineCount)) {
+            const text = document.lineAt(line).text + '\n';
+            let start = 0;
+            let end = text.length - 1;
+            let is_space = (i) => { return /\s/.test(text[i]); };
+            while ((start < text.length) && is_space(start))++start;
+            while ((end >= start) && is_space(end))--end;
+
+            return new vscode.Range(line, start, line, end);
+        } else
+            return new vscode.Range(line, 0, line, 0);
+    }
+
+    public parse(line: string): [boolean, Maybe<FileDiagnostic>] {
+        const msvc_re = /^\s*(?!\d+>)\s*([^\s>].*)\((\d+|\d+,\d+|\d+,\d+,\d+,\d+)\):\s+(error|warning|info)\s+(\w{1,2}\d+)\s*:\s*(.*)$/;
+        const res = msvc_re.exec(line);
+        if (!res)
+            return [false, null];
+        const full = res[0];
+        const file = res[1];
+        const location = res[2];
+        const severity = res[3];
+        const code = res[4];
+        const message = res[5];
+        const abspath = path.isAbsolute(file)
+            ? file
+            : path.normalize(path.join(this.binaryDir, file));
+        const loc = (() => {
+            const parts = location.split(',');
+            if (parts.length === 1)
+                this.getTrimmedLineRange(file, Number.parseInt(parts[0]) - 1);
+            if (parseFloat.length === 2)
+                return new vscode.Range(
+                    Number.parseInt(parts[0]) - 1,
+                    Number.parseInt(parts[1]) - 1,
+                    Number.parseInt(parts[0]) - 1,
+                    Number.parseInt(parts[1]) - 1
+                );
+            if (parseFloat.length === 4)
+                return new vscode.Range(
+                    Number.parseInt(parts[0]) - 1,
+                    Number.parseInt(parts[1]) - 1,
+                    Number.parseInt(parts[2]) - 1,
+                    Number.parseInt(parts[3]) - 1
+                );
+            throw new Error('Unable to determine location of MSVC error');
+        })();
+        const diag = new vscode.Diagnostic(
+            loc,
+            message,
+            {
+                error: vscode.DiagnosticSeverity.Error,
+                warning: vscode.DiagnosticSeverity.Warning,
+                info: vscode.DiagnosticSeverity.Information,
+            }[severity]
+        );
+        diag.code = code;
+        diag.source = 'MSVC';
+        return [true, {
+            filepath: abspath,
+            key: full,
+            diag: diag,
+        }];
+    }
+}
+
+/**
+ * Parses a diagnostic message from Green Hills Compiler.
+ * Use single line error reporting when invoking GHS compiler (--no_wrap_diagnostics --brief_diagnostics).
+ *
+ * @class GHSDiagnosticParser
+ * @extends {DiagnosticParser}
+ */
+class GHSDiagnosticParser extends DiagnosticParser {
+    public parse(line: string): [boolean, Maybe<FileDiagnostic>] {

```

Comment:
It seems unnecessary to return both a `boolean` and a `Maybe<...>`, as the `Maybe<T>` encodes whether something is null as part of its type. Using the `Maybe<T>` helps the compiler detect bad usages (It will produce an error if you use it without checking for `null`)